### PR TITLE
C-based harvest in ELM-FATES

### PR DESCRIPTION
**Pull request by sshu88**
This pull request contains modification in E3SM corresponding to another FATES [PR #888](https://github.com/NGEET/fates/pull/888).
In detail, necessary input to enable FATES C-based harvest from ELM is passed to FATES through `bc_in`. Output from FATES used to calculate land use emission and net biosphere productivity is passed to ELM through `bc_out`.

### CODE REVIEW

**Comment by rljacob** [comment]
The PR description should describe the code changes and not just point to another PR description.

**Comment by rljacob** [comment]
Needs action and review from @sshu88 @rgknox @bishtgautam

**Comment by rljacob** [comment]
Still needs action and review from  @sshu88 @rgknox @bishtgautam

**Comment by rljacob** [comment]
telecon notes:  Waiting for a FATES update before removing WIP.

File: components/elm/src/biogeophys/BalanceCheckMod.F90
Comment (glemieux) [inline components/elm/src/biogeophys/BalanceCheckMod.F90:622]: @sshu88 is this change to try and avoid the water balance error?  @rgknox and I recommend that we not change this in e3sm master branch.  It should be reverted to the old value.

**Review by glemieux — CHANGES_REQUESTED** [review]

File: components/elm/src/biogeophys/BalanceCheckMod.F90
Comment (rgknox) [inline components/elm/src/biogeophys/BalanceCheckMod.F90:622]: If we bump this value up to suppress warnings, that seems fine for a FATES run.  We are very aware that an audit needs to be done to reduce these fine errors in radiation scattering.  However, we may want to only apply it to a fates governed patch.  This "is_fates" can tell us if we are on a fates patch and if we should apply different logic: 
https://github.com/E3SM-Project/E3SM/blob/v2.0.0/components/elm/src/data_types/VegetationType.F90#L70

Comment (sshu88) [reply components/elm/src/biogeophys/BalanceCheckMod.F90]: I have revised the code as suggested. Thank you @rgknox @glemieux .

**Comment by rljacob** [comment]
telecon notes:  still waiting for FATES PR.

**Comment by rljacob** [comment]
@rgknox this is waiting on a FATES PR.

File: components/elm/src/dyn_subgrid/dynHarvestMod.F90
Comment (rgknox) [inline components/elm/src/dyn_subgrid/dynHarvestMod.F90:81]: @sshu88, is the long-term plan to have wood_harvest_units read in or defined in the file? or set in a namelist?

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (rgknox) [reply components/elm/src/main/elmfates_interfaceMod.F90:375]: nevermind, i see it isn't used. Can we remove this variable and its commented out usage below?

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (rgknox) [inline components/elm/src/main/elmfates_interfaceMod.F90:954]: need units for 1e6 please. I'm guessing this is m2 per km2?

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (rgknox) [inline components/elm/src/main/elmfates_interfaceMod.F90:2317]: units for 1e3 please, named constant is good too

File: components/elm/src/data_types/ColumnDataType.F90
Comment (rgknox) [inline components/elm/src/data_types/ColumnDataType.F90:5776]: remove comment

File: components/elm/src/data_types/ColumnDataType.F90
Comment (rgknox) [inline components/elm/src/data_types/ColumnDataType.F90:5798]: will NEE also be defined further down? see https://github.com/E3SM-Project/E3SM/blob/e96fc25b2d223d21530fd78ab5212feac1ecf44c/components/elm/src/data_types/ColumnDataType.F90#L6121

Comment (rgknox) [reply components/elm/src/data_types/ColumnDataType.F90:5798]: same with NEP and NBP, if already defined somewhere else that is active, lets use the same definition

**Review by rgknox — CHANGES_REQUESTED** [review]
THanks and nice work @sshu88 , I made some comments and requests. THings look pretty straightforward.

File: components/elm/src/dyn_subgrid/dynHarvestMod.F90
Comment (sshu88) [reply components/elm/src/dyn_subgrid/dynHarvestMod.F90:81]: In my opinion it would be better to set `wood_harvest_units` as a namelist option. However, this option may require the change of another set of harvest rate data in different unit.

File: components/elm/src/dyn_subgrid/dynHarvestMod.F90
Comment (sshu88) [reply components/elm/src/dyn_subgrid/dynHarvestMod.F90:185]: True. The dataset we are using now does not account for another topographic dimension.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (sshu88) [reply components/elm/src/main/elmfates_interfaceMod.F90:375]: Yes. Remove this namelist option does not affect the ELM-FATES run now.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (sshu88) [reply components/elm/src/main/elmfates_interfaceMod.F90:954]: Right.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (sshu88) [reply components/elm/src/main/elmfates_interfaceMod.F90:2317]: g per kg.

File: components/elm/src/data_types/ColumnDataType.F90
Comment (sshu88) [reply components/elm/src/data_types/ColumnDataType.F90:5798]: This repetitive definition is actually in a different IF-ELSE blocks (.not. FATES .and. C12). I'm not sure how to merge them?

Comment (rgknox) [reply components/elm/src/data_types/ColumnDataType.F90:5798]: ok, I thought that was an end if, I see its an else if now.  I have no problems with the current formulation then.  An alternative would be to move all instances of NEP, NEE and NBP to either the very beginning of the routine, or the very end, which would be outside the if fates (and wrap with ane if c12).

**Comment by bishtgautam** [comment]
@sshu88 You have mistakenly updated multiple submodules in this PR which need to be reverted.

**Comment by sshu88** [comment]
@bishtgautam @rgknox I have reverted those pointers now. Thanks.

File: components/elm/src/dyn_subgrid/dynHarvestMod.F90
Comment (aldivi) [reply components/elm/src/dyn_subgrid/dynHarvestMod.F90:81]: I agree that a namelist option would be best in the long-term, particularly if the input file eventually contains both carbon and area prescriptions.

**Comment by bishtgautam** [comment]
@sshu88 Thanks for fixing the submodules. Is this PR still WIP or ready to be reviewed/merged or waiting for another FATES PR to go first?

cc: @glemieux @rgknox

File: components/elm/src/data_types/ColumnDataType.F90
Comment (bishtgautam) [inline components/elm/src/data_types/ColumnDataType.F90:6988]: Please undo the deletion of space

**Review by bishtgautam — COMMENTED** [review]
I have requested minor changes.

**Review by rgknox — APPROVED** [review]

**Comment by bishtgautam** [comment]
@sshu88 Is this WIP or ready to be merged?

**Comment by sshu88** [comment]
@bishtgautam We can go ahead and merge the work once the FATES part is merged.

**Comment by glemieux** [comment]
> @bishtgautam We can go ahead and merge the work once the FATES part is merged.

@bishtgautam I'm working on the final deconflict and test on the fates-side right now.

**Comment by rgknox** [comment]
The changes just merged into this branch, in https://github.com/E3SM-Project/E3SM/pull/5106/commits/13d0f5a71ebff055ddd3155a13ebcdd69da43ce7 , were originally overlooked, but nonetheless required for compatibility with fates API 25.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (glemieux) [inline components/elm/src/main/elmfates_interfaceMod.F90:488]: I don't see any setup for this on the fates side pr 888.  Is this a future feature?  I'm assuming so given that it's hard coded to false just above this.

**Review by glemieux — COMMENTED** [review]
One question about the hardcoded `allow_harvest_bypass_criteria`

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (sshu88) [reply components/elm/src/main/elmfates_interfaceMod.F90:488]: It's an old flag. I forgot to delete this line in ELM side.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (glemieux) [inline components/elm/src/main/elmfates_interfaceMod.F90:160]: @sshu88 it looks like there is some duplicate code here as well.

**Review by glemieux — APPROVED** [review]
Just one more minor piece of code cleanup.

File: components/elm/src/main/elmfates_interfaceMod.F90
Comment (sshu88) [reply components/elm/src/main/elmfates_interfaceMod.F90:160]: Thanks for checking.

**Comment by glemieux** [comment]
@bishtgautam we still need to update the pointer to the fates side, but the associated PR hasn't been merged yet.  I'll give you a heads up when that happens

**Comment by bishtgautam** [comment]
Are the existing tests cover the new physics? Or, should we add a new test?

**Comment by glemieux** [comment]
> Are the existing tests cover the new physics? Or, should we add a new test?

We should add new tests, but I would like to defer that to a future PR in which we review and update all of the fates' tests for the land developer suite.

File: components/elm/src/data_types/ColumnDataType.F90
Comment (glemieux) [inline components/elm/src/data_types/ColumnDataType.F90:2102]: @sshu88 was this a temporary change to make sure that you could read the `prod10c` and `prod100c` for when fates was running?

**Review by glemieux — COMMENTED** [review]
I noticed one change that affects non-fates runs that needs to be addressed.

File: components/elm/src/data_types/ColumnDataType.F90
Comment (sshu88) [reply components/elm/src/data_types/ColumnDataType.F90:2102]: These are two outputs I added to the history output file for checking the change of product pool size. So it is a temporary change. But the product pool size shall be a basic output in land surface model when considering anthropogenic disturbances.

Comment (glemieux) [reply components/elm/src/data_types/ColumnDataType.F90:2102]: To clarify, does this need to be available when fates is running?  The current issue is that the same call is made just below behind a `.not. use_fates` check, which means this will get called twice and cause a build failure for non-fates runs.  If it's not necessary for fates right now, would you mind removing these lines?  Otherwise, the part that is behind the `.not. use_fates` check could be removed.

Comment (sshu88) [reply components/elm/src/data_types/ColumnDataType.F90:2102]: I have removed the part below `.not. use_fates` check.

**Comment by glemieux** [comment]
I'm running the land-developer tests and seeing two unexpected failures.  I haven't looked into them much yet, but one is a fates smoke test failing during runtime in PhosphorusDynamicsMod and the other is a model build failure for `ERS.ne30pg2_r05_EC30to60E2r2.GPMPAS-JRA.cori-haswell_intel.mosart-rof_ocn_2way`.

folder location (Cori): 
```
/global/homes/g/glemieux/cori-scratch/e3sm-tests/pr5106-merge.fates.cori-haswell.intel.Ce0716a090f-F25047167
```

Note I'm using Cori as perlmutter is down right now.

**Comment by rgknox** [comment]
I submitted some updates to @sshu88 's branch to resolve the issues with the fates_eca test.  Tests were passing on Cori

**Comment by glemieux** [comment]
@sshu88 I created a [PR to your branch](https://github.com/sshu88/E3SM/pull/6) to update the fates pointer now that https://github.com/NGEET/fates/pull/888 is integrated on the fates-side.

@bishtgautam I ran land developer tests on Cori and all the tests except `ERS.ne30pg2_r05_EC30to60E2r2.GPMPAS-JRA.gcp12_gnu.mosart-rof_ocn_2way` passed (due to https://github.com/E3SM-Project/E3SM/issues/5412).

**Comment by glemieux** [comment]
@bishtgautam this shoudl be closed per #5429.

**Pull request closed without merging**
